pkg/progressbar/progressbar.go

```diff
@@ -9,6 +10,17 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+type ProxyReaderAt struct {
+	io.ReaderAt
+	Bar *pb.ProgressBar
+}
+
+func (r *ProxyReaderAt) ReadAt(p []byte, off int64) (int, error) {
+	n, err := r.ReaderAt.ReadAt(p, off)
+	r.Bar.Add(n)
```

Comment:
This should check negative n?

Reply:
This is documented to return positive value so I think we are safe.

> WriteAt writes len(p) bytes from p to the underlying data stream at offset off. It returns the number of bytes written from p (0 <= n <= len(p)) and any error encountered that caused the write to stop early. WriteAt must return a non-nil error if it returns n < len(p).

If we get negative value the convert will fail here:
https://github.com/lima-vm/go-qcow2reader/pull/36/commits/2cdb2331b6edce792b6d92c1fa35cb8e09736fe5#diff-b4093fac13dc66eefdab36a0bd2f4ee72c635a84a9c5687f8fc3e33185c6d9aeR131

Reply:
If err is non-nil, r.Bad.Add should not be called?

Reply:
> If err is non-nil and n is negative

n is never negative

I pasted io.WriterAt docs before, but io.ReaderAt docs are the same:

> ReadAt reads len(p) bytes into p starting at offset off in the underlying input source. It returns the number of bytes read (0 <= n <= len(p)) and any error encountered.

> r.Bar.Add should not be called?

err may be io.EOF:

> If the n = len(p) bytes returned by ReadAt are at the end of the input source, ReadAt may return either err == EOF or err == nil.

So I think always adding n and let the caller handle errors in the right thing. This is also what pb.Reader is doing: https://github.com/cheggaaa/pb/blob/4ca3463f4bcf4446e1746efc6e484e795cfe7b12/reader.go#L15

We can handle n == 0, but I'm not sure it is helpful to add a check that fail for all good reads, for case that happens only when err != nil, maybe once when per convert.